ENH: Adding origin parameter in pd.to_datetime

sumitbinnani opened the pull request:
Fixes #11276
Fixes  #11745

jreback commented:
tests!

sumitbinnani commented:
On it! Any boundary conditions that I must take into consideration?

jreback commented:
what do you mean?

jreback commented:
also this would need to be added in Timestamp as that accepts a unit as well

sumitbinnani commented:
Apologies for the late reply. I am not able to understand which file to edit for Timestamp. Kindly guide, @jreback?

jreback commented:
https://github.com/pydata/pandas/blob/master/pandas/tslib.pyx#L222

sumitbinnani commented:
Thanks @jreback. Will go through the file asap :-)

sumitbinnani commented:
Hi, I am bit confused regarding the implementation. As far as I could understand, following might help. Kindly correct me if I am wrong.

```
## Constructor in Timestamp Class.
def __new__(cls, object ts_input, object offset=None, tz=None, unit=None):
        cdef _TSObject ts
        cdef _Timestamp ts_base

        if unit == 'julian':
            unit = 'D'
            ts_input = ts_input - Timestamp(0).to_julian_date()

        ts = convert_to_tsobject(ts_input, tz, unit)

        if ts.value == NPY_NAT:
            return NaT

        if util.is_string_object(offset):
            from pandas.tseries.frequencies import to_offset
            offset = to_offset(offset)

        # make datetime happy
        ts_base = _Timestamp.__new__(cls, ts.dts.year, ts.dts.month,
                                     ts.dts.day, ts.dts.hour, ts.dts.min,
                                     ts.dts.sec, ts.dts.us, ts.tzinfo)

        # fill out rest of data
        ts_base.value = ts.value
        ts_base.offset = offset
        ts_base.nanosecond = ts.dts.ps / 1000

        return ts_base
```

Does it seem okay, @jreback?

jreback commented:
@sumitbinnani this should be implemented in ``cast_from_unit``

jreback commented:
this needs tests

jreback commented:
see #11745; let's call this ``origin`` (and in this case I think it would be a string, e.g. 'julian', rather than a specific date)

sumitbinnani commented:
Have been a bit occupied for last a few weeks owing to applications and submission deadlines. Will make the necessary changes for #11745 asap. Facing a bit difficulty writing test cases (going through documentations and will mostly be done by this weekend).

sumitbinnani commented:
Have changes the definition of  ``to_datetime`` to include a new parameter ``origin``, and have added following lines in the code:
```
    ...
    if origin is not None:
        import datetime
        from pandas.core.api import Timestamp
        if origin == 'julian':
            arg = arg - Timestamp(0).to_julian_date()
        elif isinstance(origin, datetime.date):
            arg = arg + Timestamp(origin).to_julian_date() - Timestamp(0).to_julian_date()
   
   return ...
```

Should I remove the changes made to the class ``Timestamp`` as ``unit`` is no longer being used for the required functionality?

jreback commented:
@sumitbinnani yes, remove the original ``unit`` changes and go with ``origin`` as discussed

sumitbinnani commented:
Hi @jreback, 

Have made the changes as followed:
* Added ``origin`` as a new parameter with default value of ``None`` to ``to_datetime``
* Updated documentation for ``to_datetime``
* Added test case for the patch (wrote test case for the first time and might require a bit of changes)

Let me know if any additional changes are required.

jreback commented on pandas/tseries/tests/test_timeseries.py:
```diff
@@ -743,6 +743,22 @@ def test_to_datetime_unit(self):
         assert_series_equal(result,expected)
+        
+    def test_to_datetime_origin(self):
+        # for origin as julian
```
add the issue number as a comment

jreback commented on pandas/tseries/tools.py:
```diff
@@ -224,6 +224,8 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
     infer_datetime_format : boolean, default False
         If no `format` is given, try to infer the format based on the first
         datetime string. Provides a large speed-up in many cases.
+    origin : datetime.date, default None
```
```
origin : scalar convertible to Timestamp
    defines the relative offset for the returned dates
```

jreback replied:
add an example of why you would want to do this (you can add in the ``Examples`` section).

jreback commented on pandas/tseries/tests/test_timeseries.py:
```diff
@@ -1,6 +1,6 @@
 # pylint: disable-msg=E1101,W0612
 import calendar
-from datetime import datetime, time, timedelta
+from datetime import date, datetime, time, timedelta
```
we don't use ``datetime.date``

jreback commented on pandas/tseries/tools.py:
```diff
@@ -271,6 +273,15 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
+    
+    if origin is not None:
+        import datetime
+        from pandas.core.api import Timestamp
```
don't need any imports here
```
if origin == 'julian':
    ....
elif if origin is not None:
   arg = arg - Timestamp(origin)
```

jreback commented on pandas/tseries/tests/test_timeseries.py:
```diff
@@ -743,6 +743,22 @@ def test_to_datetime_unit(self):
+        result = Series(pd.to_datetime(daysFromEpoch, unit='D', origin=EPOCH1960))
+        expected = Series(pd.date_range('1960-1-1', periods=5))
+        assert_series_equal(result,expected)
+        
```
test with origin as something else as well (e.g. a different string/timestamp convertible), use a loop to pass ``date,datetime,Timestamp,np.datetime64,string`` types (these are all ``Timestamp`` convertible).

jreback commented:
pls add whatsnew note in Enhancements for 0.18.0

sumitbinnani commented:
Have incorporated the changes in Code Review.

The file for what's new in 0.18.0, is in commit ahead of my branch's head. The rebase command isn't working on my system (maybe configuration issue). Should I copy the file from master branch and make changes, or should I submit a new pull requests for the same?

jorisvandenbossche commented on pandas/tseries/tools.py:
```diff
@@ -263,6 +266,17 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
     99   2000-04-09
     Length: 100, dtype: datetime64[ns]
+    
+    Getting datetime as number od days since 1/1/1960
```
od -> of

sumitbinnani replied:
Rectified.

jorisvandenbossche commented on pandas/tseries/tools.py:
```diff
@@ -224,6 +224,9 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
         datetime string. Provides a large speed-up in many cases.
+    origin : scalar convertible to Timestamp / `julian`
+        Define relative offset for the returned dates
+        - If `julian`, offset is set to beginning of Julian Calendar
```
Can you let the first sentence ('Define ...') end with a `.`, and remove the `-` on this line? As this will not be rendered as a list in this way, but just as continuous text (which is OK here)

sumitbinnani replied:
Rectified.

jorisvandenbossche commented:
Can you just try:

```
git fetch upstream
git rebase -i upstream/master
```

Does this not work?

jreback commented on pandas/tseries/tools.py:
```diff
@@ -263,6 +266,17 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
+    
+    >>> EPOCH1960 = date(1960, 1, 1)
+    >>> daysFromEpoch = range(100)
+    >>> pd.to_datetime(daysFromEpoch, unit='D', origin=EPOCH1960)
```
just do this in a single line, no need for the temporaries

sumitbinnani replied:
Rectified.

jreback commented on pandas/tseries/tools.py:
```diff
@@ -271,6 +285,14 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
     """
+    
+    if origin is not None:
+        from pandas.core.api import Timestamp
```
already imported

sumitbinnani replied:
Rectified.

jreback commented on pandas/tseries/tools.py:
```diff
@@ -271,6 +285,14 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
+        if origin == 'julian':
+            arg = arg - Timestamp(0).to_julian_date()
+        else:
+            arg = arg + Timestamp(origin).to_julian_date() - Timestamp(0).to_julian_date()
```
the else just needs to be:

``arg = arg - Timestamp(origin)``

sumitbinnani replied:
Since I am subtracting from the raw input, I would have to also account for the default offset behavior. i.e.
```
arg = arg - (defaultOrigin - newOrigin)
```

jreback replied:
I think we need some examples for this use case and tests for the same.

sumitbinnani replied:
Have already added a few tests case.
https://github.com/sumitbinnani/pandas/blob/9949933259f28e59b5777cf8bc46494dd29d1556/pandas/tseries/tests/test_timeseries.py#L755

Do we need a few more?

jreback replied:
I want to see some more actual usecases. All of the tests are just variations on a single date.

sumitbinnani replied:
Would add a few more cases. Anything in particular that you think should be included.

jreback replied:
maybe look and see what examples R uses. This is where this came from AFAIK

jreback commented on pandas/tseries/tools.py:
```diff
@@ -224,6 +224,9 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
     infer_datetime_format : boolean, default False
         If no `format` is given, try to infer the format based on the first
         datetime string. Provides a large speed-up in many cases.
+    origin : scalar convertible to Timestamp / `julian`
```
or ``string 'julian'``

sumitbinnani replied:
Rectified.

jreback commented on pandas/tseries/tools.py:
```diff
@@ -224,6 +224,9 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
+        Define relative offset for the returned dates
+        - If `julian`, offset is set to beginning of Julian Calendar
 
     Returns
```
set what origin means if its not 'julian'

sumitbinnani replied:
Rectified.

jreback commented:
pls add a whatsnew entry (in enhancements)

sumitbinnani commented:
@jorisvandenbossche, I am getting the following error:
```
git: 'rebase' is not a git command. See 'git --help'.
```

The error started popping a week back, and I guess is due to the configuration problem. Have even tried reinstalling git, but the error still exists. Should I simply copy paste the enhancement file from master branch and make the necessary edits? Or should I submit a small patch for the same?

jorisvandenbossche commented:
@sumitbinnani I would rather try to fix the git command (you will need it more). Maybe look at http://stackoverflow.com/questions/10125113/git-rebase-is-not-a-git-command-see-git-help

sumitbinnani commented:
@jorisvandenbossche I am planning to reinstall my OS, as a few more applications/commands are not working correctly after I patched a few packages. However, I am reluctant to do so in hurry owing to data loss, and massive time to get all the packages and libraries required for my job and day to day activities.

Have worked upon the code review and will push the changes after adding more test-cases to avoid multiple commits (given my problem with rebase command).

jreback commented:
@sumitbinnani you should just use conda, this makes working with pandas (and all other python packages) very very easy. see the contributing page http://pandas.pydata.org/pandas-docs/version/0.17.1/contributing.html

sumitbinnani commented:
Hi @jreback, I am using conda only and I am having no issues with python packages and libraries. They are working like a charm. But a few other installation like git have gone haywire and are giving error on otherwise valid commands. I would look into the issue during the long weekend about to come. Meanwhile, I will submit a small patch for the edit in enhancement file.

sumitbinnani commented:
@jreback: Even the documentation in R have a single use case for the ``origin``. However, they have documented a bit about the possible scenarios where it can be used.
ref: https://stat.ethz.ch/R-manual/R-devel/library/base/html/as.Date.html

sumitbinnani commented:
Does the patch requires any more changes? @jreback? @jorisvandenbossche?

shoyer commented:
Maybe it would better to have the default be `origin='unix'`, to more explicitly indicate that time is counted from the unix epoch?

shoyer commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -33,6 +33,7 @@ Other enhancements
 
 - Handle truncated floats in SAS xport files (:issue:`11713`)
 - ``read_excel`` now supports s3 urls of the format ``s3://bucketname/filename`` (:issue:`11447`)
+- ``to_datetime`` have a new parameter, ``origin``, to define offset for ``DateTimeIndex`` (:issue:`11276`)
```
should be `DatetimeIndex`

shoyer commented:
It looks like the current implementation is broken unless `unit='D'`. Please add tests and any necessary fixes for other time units.

sumitbinnani commented:
@shoyer The current implementation is indeed breaking unless ``unit='D'``. Would resolve it asap.

jreback commented on pandas/tseries/tools.py:
```diff
@@ -224,6 +224,10 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
     infer_datetime_format : boolean, default False
         If no `format` is given, try to infer the format based on the first
         datetime string. Provides a large speed-up in many cases.
+    origin : scalar convertible to Timestamp / string 'julian', default None
```
I agree with @shoyer here, let's define the default of ``origin='unix'`` (and allow only ``julian``), other values are errors.

sumitbinnani replied:
Sure. Would keep default as ``unix``.  A few of the cases for the use of arbitrary dates have been mentioned [here](https://stat.ethz.ch/R-manual/R-devel/library/base/html/as.Date.html).

shoyer replied:
Yes, there are definitely good use cases for arbitrary origins.

On Tue, Dec 29, 2015 at 11:06 PM, Sumit Binnani <notifications@github.com>
wrote:

>> @@ -224,6 +224,10 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
>>      infer_datetime_format : boolean, default False
>>          If no `format` is given, try to infer the format based on the first
>>          datetime string. Provides a large speed-up in many cases.
>> +    origin : scalar convertible to Timestamp / string 'julian', default None
> Sure. Would keep default as ``unix``.  A few of the cases for the use of arbitrary dates have been mentioned [here](https://stat.ethz.ch/R-manual/R-devel/library/base/html/as.Date.html).
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/pydata/pandas/pull/11470/files#r48589485

jreback replied:
do you have some other examples / links that we can add here?

sumitbinnani replied:
Would search a bit and update.

sumitbinnani replied:
This [Wikipedia Article](https://en.wikipedia.org/wiki/Epoch_(reference_date)#Notable_epoch_dates_in_computing) seems like a good way to exmplify possible use cases.

jreback commented:
can you rebase / update

jreback commented:
@sumitbinnani can you rebase/update

sumitbinnani commented:
Sorry for the late response. I was a bit occupied with applications and hence the delay. I am left with a small issue for ``unit='ms'``. Would work upon it and update by end of week.

sumitbinnani commented:
Any way to generate random test cases for different values of parameter ``unit``?

jorisvandenbossche commented:
Why do you want to generate random test cases? Just test some different values?

sumitbinnani commented:
@jorisvandenbossche The way I am generating test cases, speaking naively, is by reverse engineering the code. Hence, thought to make it random for robustness.

jreback commented:
can you rebase/update

sumitbinnani commented:
Hi, There are some issues while changing origin. The current code is goes as follow:

```
def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
                utc=None, box=True, format=None, exact=True, coerce=None,
                unit='ns', infer_datetime_format=False, origin='unix'):

    offset = 0
    if origin == 'unix':
        return _to_datetime(arg, errors=errors, dayfirst=dayfirst, yearfirst=yearfirst,
                        utc=utc, box=box, format=format, exact=exact,
                        unit=unit, infer_datetime_format=infer_datetime_format)
    elif origin == 'julian':
        offset = to_datetime(0)-to_datetime(0)
    else:
        offset = tslib.Timestamp(origin)
        
    return _to_datetime(arg, errors=errors, dayfirst=dayfirst, yearfirst=yearfirst,
                        utc=utc, box=box, format=format, exact=exact,
                        unit=unit, infer_datetime_format=infer_datetime_format) - tslib.Timestamp(0) + offset

```

But it returns, timedelta instead of datetimeindex in a few scenarios. Have tried multiple combinations, but none had been working. Any idea on how to proceed?

jreback commented:
can you show which scenarios (e.g. input args)?

```
In [35]: pd.to_datetime(0)-pd.to_datetime(0)
Out[35]: Timedelta('0 days 00:00:00')
```

subtracting 2 datetimes is by definition a ``Timedelta``

sumitbinnani commented:
The problem is occurring in case the origin is defined as ``julian``.

jreback commented:
@sumitbinnani you need to rebase on master. 

```git rebase -i master``

see [here](http://pandas.pydata.org/pandas-docs/stable/contributing.html#combining-commits)

sumitbinnani commented:
The branch have been rebased on master and conflicts have been resolved. Does it look okay, @jreback?

jreback commented:
not sure why you have so many commits. pls squash it down.

sumitbinnani commented:
Have squashed the commits. Does it look good, @jreback?

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -522,6 +522,7 @@ Other enhancements
 - ``HDFStore`` is now iterable: ``for k in store`` is equivalent to ``for k in store.keys()`` (:issue:`12221`).
 - Add missing methods/fields to ``.dt`` for ``Period`` (:issue:`8848`)
 - The entire codebase has been ``PEP``-ified (:issue:`12096`)
```
move to 0.18.1

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -522,6 +522,7 @@ Other enhancements
 - ``HDFStore`` is now iterable: ``for k in store`` is equivalent to ``for k in store.keys()`` (:issue:`12221`).
 - Add missing methods/fields to ``.dt`` for ``Period`` (:issue:`8848`)
 - The entire codebase has been ``PEP``-ified (:issue:`12096`)
+- ``to_datetime`` have a new parameter, ``origin``, to define offset for ``DatetimeIndex`` (:issue:`11276`)
```
going to need a mini-example

jreback commented on pandas/tseries/tests/test_timeseries.py:
```diff
@@ -757,6 +757,26 @@ def test_to_datetime_unit(self):
+        # for origin as julian
+        julian_dates = pd.date_range('2014-1-1', periods=10).to_julian_date().values
+        result = Series(pd.to_datetime(julian_dates, unit='D', origin='julian'))
+        expected = Series(pd.to_datetime(julian_dates - pd.Timestamp(0).to_julian_date(), unit='D'))
```
don't convert to series, leave as index and use ``assert_index_equal``

sumitbinnani replied:
This gives assertion error.

```
AssertionError: [index] Expected object to be of type <class 'pandas.indexes.base.Index'>, found <type 'list'> instead
```

jreback commented on pandas/tseries/tests/test_timeseries.py:
```diff
@@ -757,6 +757,26 @@ def test_to_datetime_unit(self):
+            for unit in units:
+                expected = Series([pd.Timedelta(x,unit=unit)+EPOCH1960 for x in unitsFromEPOCH])
+                result = Series(pd.to_datetime(unitsFromEPOCH, unit=unit, origin=EPOCH))
+                assert_series_equal(result, expected)
```
need tests for error conditions. E.g. what if origin is a ``string``, ``datetime``, integer. when should it raise.

sumitbinnani replied:
Have already accounted for origin for types ``string``, ``datetime``, ``datetime64``

jreback commented on pandas/tseries/tools.py:
```diff
@@ -222,6 +222,11 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
         datetime strings, and if it can be inferred, switch to a faster
         method of parsing them. In some cases this can increase the parsing
         speed by ~5-10x.
+    origin : scalar convertible to Timestamp / string ('julian', 'unix'), default 'unix'
```
let's call this default of ``epoch`` (for ``unix``)

sumitbinnani replied:
Rectified.

jreback commented on pandas/tseries/tools.py:
```diff
@@ -222,6 +222,11 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
+    origin : scalar convertible to Timestamp / string ('julian', 'unix'), default 'unix'
+        Define relative offset for the returned dates.
+        - If 'unix', offset is set to 1-1-1970
+        - If 'julian', unit must be 'D', and offset is set to beginning of Julian Calendar
```
I think ``unit`` must be ``None`` (or maybe ``D``). need to test on these and other conditions (which should error)

sumitbinnani replied:
Googled a bit. The julian dates are always define in unit of days.

jreback commented on pandas/tseries/tools.py:
```diff
@@ -222,6 +222,11 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
+        Define relative offset for the returned dates.
+        - If 'unix', offset is set to 1-1-1970
+        - If 'julian', unit must be 'D', and offset is set to beginning of Julian Calendar
+        - If Timestamp convertible, offset is set to Timestamp identified by origin
```
are we requiring a normalized Timestamp? IOW, no hhmmss......? (I think yes, so ``pd.Timestamp(julian) == pd.Timestamp(julian).normalize()`` or its an error.

sumitbinnani replied:
Unable to comprehend :-(

jreback commented on pandas/tseries/tools.py:
```diff
@@ -275,6 +280,15 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
 
+    Getting datetime as number of days since 1/1/1960
+
+    >>> pd.to_datetime(range(100), unit='D', origin=date(1960, 1, 1))
```
use ``Timestamp('1960-01-01')``, ``date`` is not commonly used.

jreback commented on pandas/tseries/tools.py:
```diff
@@ -283,10 +297,23 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
+                            unit=unit, infer_datetime_format=infer_datetime_format)
+    elif origin == 'julian':
+        if unit != 'D':
+            raise AssertionError("unit must be 'D' for origin='julian'")
```
raise ``ValueErrors``

jreback commented on pandas/tseries/tools.py:
```diff
@@ -283,10 +297,23 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
+        offset = tslib.Timestamp(origin) - tslib.Timestamp(0)
+        return _to_datetime(arg, errors=errors, dayfirst=dayfirst, yearfirst=yearfirst,
+                            utc=utc, box=box, format=format, exact=exact,
+                            unit=unit, infer_datetime_format=infer_datetime_format) + offset
```
instead of repeating all this code, just do:

```
offset = None
if origin == 'epoch':
    pass
elif origin == 'julian':
    # check errors
    arg = arg - Timestamp(0).to_julian_date()    
else:
    offset = Timestamp(origin) - Timestamp(0)

result = _to_datetime(.......)
if offset is not None:
     result = result + offset
return result
```

sumitbinnani commented:
@jreback Have worked on code review and committed the code.

jreback commented on pandas/tseries/tools.py:
```diff
@@ -297,23 +297,22 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
-    elif origin == 'julian':
+    offset = None
+    if origin == 'julian':
         if unit != 'D':
```
you can also accept ``unit is None`` here

sumitbinnani replied:
The default for ``unit`` is ``ns``. So explicitly providing it as a ``None`` might be confusing (as julian dates are always defines as 'julian day'). Let me know what do you think, @jreback?

sumitbinnani commented:
The default for unit is ns. So explicitly providing it as a None might be confusing (as julian dates are always defines as 'julian day'). Let me know what do you think, @jreback?

jreback commented on pandas/tseries/tests/test_timeseries.py:
```diff
@@ -757,6 +757,26 @@ def test_to_datetime_unit(self):
+                assert_series_equal(result, expected)
+
     def test_series_ctor_datetime64(self):
         rng = date_range('1/1/2000 00:00:00', '1/1/2000 1:59:50', freq='10s')
```
need some tests for invalid ``origin`` and invalid combination of ``origin='julian'`` and ``unit != D``

sumitbinnani commented:
Have added the test cases for invalid ``origin`` and invalid combination of ``origin='julian'`` and ``unit != 'D'``. Does it look good, @jreback?

jorisvandenbossche commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -59,6 +59,7 @@ Other Enhancements
 - ``pd.read_csv()`` now supports opening ZIP files that contains a single CSV, via extension inference or explict ``compression='zip'`` (:issue:`12175`)
 - ``pd.read_csv()`` now supports opening files using xz compression, via extension inference or explicit ``compression='xz'`` is specified; ``xz`` compressions is also supported by ``DataFrame.to_csv`` in the same way (:issue:`11852`)
 - ``pd.read_msgpack()`` now always gives writeable ndarrays even when compression is used (:issue:`12359`).
+- ``pd.to_datetime`` have a new parameter, ``origin``, to define offset for ``DatetimeIndex`` (:issue:`11276`)
```
have -> has

jorisvandenbossche commented on pandas/tseries/tools.py:
```diff
@@ -283,10 +297,22 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
+        offset = tslib.Timestamp(origin) - tslib.Timestamp(0)
+
+    result = _to_datetime(arg, errors=errors, dayfirst=dayfirst, yearfirst=yearfirst,
+                            utc=utc, box=box, format=format, exact=exact,
```
Can you fix the indentation here? (two space of with 'arg')

jorisvandenbossche commented on pandas/tseries/tools.py:
```diff
@@ -222,6 +222,11 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
         speed by ~5-10x.
+    origin : scalar convertible to Timestamp / string ('julian', 'epoch'), default 'epoch'
+        Define relative offset for the returned dates.
+        - If 'epoch', offset is set to 1-1-1970
```
Can you leave a blank line between "Define ..." and "- If .." (needed for correct formatting of the restructured text)

jorisvandenbossche commented:
From looking at the implementation, I suspect that using a custom ``origin`` will always work, also when the values to parse are already dates (eg `to_datetime([2012-01-01'], origin='2000-01-01')`). This is a bit of a cornercase and one should never do that, but I think we should raise an error in such circumstances instead of silently returning garbage? 

Is it correct to state that `origin` only makes sense when dealing with integer values that are interpreted based on the `unit` kwarg?

sumitbinnani commented:
Have worked upon the code review and push the changes.

@jorisvandenbossche: I couldn't think of any probable use case. Useful only when I need to find new dates assuming that initial epoch provided was incorrect (tweaking with data, extremely rare case). Maybe we could raise a warning, @jreback?

jreback commented on pandas/tseries/tools.py:
```diff
@@ -222,6 +222,12 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
+    origin : scalar convertible to Timestamp / string ('julian', 'epoch'), default 'epoch'
+        Define relative offset for the returned dates.
+        
+        - If 'epoch', offset is set to 1-1-1970
```
add a versionadded tag here

jreback commented on pandas/tseries/tools.py:
```diff
@@ -283,10 +298,22 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
-                        yearfirst=yearfirst,
-                        utc=utc, box=box, format=format, exact=exact,
-                        unit=unit, infer_datetime_format=infer_datetime_format)
+    offset = None
```
give a 1-liner about what this is doing here

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -59,6 +59,7 @@ Other Enhancements
 - ``pd.read_csv()`` now supports opening ZIP files that contains a single CSV, via extension inference or explict ``compression='zip'`` (:issue:`12175`)
 - ``pd.read_csv()`` now supports opening files using xz compression, via extension inference or explicit ``compression='xz'`` is specified; ``xz`` compressions is also supported by ``DataFrame.to_csv`` in the same way (:issue:`11852`)
 - ``pd.read_msgpack()`` now always gives writeable ndarrays even when compression is used (:issue:`12359`).
+- ``pd.to_datetime`` has a new parameter, ``origin``, to define offset for ``DatetimeIndex`` (:issue:`11276`)
```
add also 11745 as well

jreback replied:
to define an offset for

jreback commented:
ok, pls rebase. then run

``git diff master | flake8 --diff``. after you have made these corrections, ping when green.

sumitbinnani commented:
Checks passed. Branch rebased and no more conflicts. Hope everything is as required, @jreback?

jreback commented:
```
In [2]: pd.to_datetime([1,2],unit='D',origin='julian')
OverflowError: Python int too large to convert to C long
```
This seems wrong

jreback commented:
pls squash as well

sumitbinnani commented:
@jreback: the error occurs because the julian dates corresponding to 1,2 is out of the range that could be handled by pandas. Similar error also occurs when you try to set origin very high or low. e.g. ``origin='1001-01-01'``.

jreback commented:
@sumitbinnani that is not a very nice error, it should do something intelligent or a better error message.

jreback commented on pandas/tseries/tools.py:
```diff
@@ -283,10 +303,25 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
+    if origin == 'julian':
+        if unit != 'D':
+            raise ValueError("unit must be 'D' for origin='julian'")
+        arg = arg - tslib.Timestamp(0).to_julian_date()
```
subtracting the arg may not make sense here, in fact shouldn't it be converted to a julian date?

need some more testing with ``origin='julian'``

```
In [30]: pd.to_datetime(pd.to_datetime([1,2],unit='D').to_julian_date()-pd.Timestamp(0).to_julian_date(),unit='D')
Out[30]: DatetimeIndex(['1970-01-02', '1970-01-03'], dtype='datetime64[ns]', freq=None)
```

sumitbinnani replied:
The ``to_julian_date`` method is an example of [Variant 1 with epoch: 12:00PM Jan 1, 4713 BC](https://en.wikipedia.org/wiki/Julian_day#Variants). Thus the desired output for input ``[1,2]`` is ``['4713BC-01-02 12:00PM','4713BC-01-03 12:00PM' ]`` which is out of the the range that could be handled.

sumitbinnani replied:
Online implementation of variant being used, [link](https://www.aavso.org/jd-calculator).

jreback replied:
I c. ok

then we should catch these out-of-ranges and either return actual datetimes or raise a nicer error message (I think the later is simply better)

```
In [1]: pd.to_datetime(['1000-01-01'])
OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 1000-01-01 00:00:00
```

Furthermore should be able to handle this situation (and need tests).
```
In [2]: pd.to_datetime(['1000-01-01'],errors='coerce')
Out[2]: DatetimeIndex(['NaT'], dtype='datetime64[ns]', freq=None)
```

sumitbinnani replied:
I am trying to understand the current methodology of raising the error. Would mostly be able to resolve the same by end of week. Thanks,

jreback replied:
I don't this is valid, you haven't asserted (or converted) arg at this point, how can you subtract?

jreback replied:
I believe that arg MUST be integer/float at this point, right? we need to check that and/or use a try/except. and a test as well.

jorisvandenbossche replied:
I think you can do this simpler, and check for the origin keyword before the actual conversion (the code you now put in `result_without_offset`. Something like 

```
if unit is not None and offset != 'epoch':
    if origin == 'julian':
        .... (check for unit == D and define offset)
    else:
         offset = ...

    arg += offset

jreback commented:
lmk when you have updated

sumitbinnani commented:
Hi @jreback, I could think of three places to raise the error:

1.  in ``tools.py: to_datetime`` function
2.  in ``tslib.pyx: array_to_datetime``
3.  in ``tslib.pyx: cast_from_unit`` (this is where the actual error occurs)

Let me know where do you think that it would be apt to raise the error?

sumitbinnani commented:
I guess raising it in ``cast_from_unit`` should be a better option`as it would be more generic and can be useful for other cases also.

jreback commented:
you just need to catch the error and raise something more useful
don't change anything lower level

jreback commented:
@sumitbinnani pls rebase/update.

jorisvandenbossche commented:
Can you also add a check for my comment above (https://github.com/pydata/pandas/pull/11470#issuecomment-211921005). 
I think we should raise an error when `origin` is used when not passing integers.

sumitbinnani commented:
@jorisvandenbossche: Sure... but there are some cases where it might be useful (we stored data with wrong epoch).

jorisvandenbossche commented:
But if you have a full series of timestamps created with a wrong epoch, shouldn't you just do `s + pd.Timedelta(..)` instead of using `to_datetime`?

sumitbinnani commented:
@jorisvandenbossche: would make necessary changes. Wasn't able to think that through.

jreback commented:
@sumitbinnani if you can rebase and make the changes ASAP. ty.

sumitbinnani commented:
@jreback: Facing difficulty with respect to error handling as a few cases haven't been handled in the existing version. See #13030.

jreback commented:
go ahead and add whatever test u need assuming the other issue is fixed - u can comment it out or just assert things for now

jreback commented:
ok I merged #13033 
pls rebase on master and update as soon as possible.

jreback commented:
Ok after my patch, these now have meaningful errors.

```
In [5]: pd.to_datetime([1,2],unit='D',origin='julian')
ValueError: cannot convert input with unit: D

In [6]: pd.to_datetime([1,2],unit='D',origin='julian',errors='coerce')
Out[6]: DatetimeIndex(['NaT', 'NaT'], dtype='datetime64[ns]', freq=None)

In [7]: pd.to_datetime([1,2],unit='D',origin='julian',errors='ignore')
Out[7]: Index([-2440586.5, -2440585.5], dtype='object')
```

I don't think you need to actually change any code except for adding some tests which assert these.

(aside from @jorisvandenbossche change).

for [5] you may want to catch this and raise with an even more meaningful message.

jreback commented:
I fixed this up here: https://github.com/jreback/pandas/commit/0b8368daa5ff2d3de283eeb8601b896779e5ebbc

I think I introduced a bug when parsing valid units with NaT's that might be strings. So fixed here as well.

@jorisvandenbossche what was your comment above referring to?

jorisvandenbossche commented:
@jreback which comment do you mean?

jorisvandenbossche commented:
> ```
In [5]: pd.to_datetime([1,2],unit='D',origin='julian')
ValueError: cannot convert input with unit: D
```

@jreback why is 'unit D' not valid? You get a out of bounds date here, but that is not really clear from that message.

jreback commented:
https://github.com/pydata/pandas/pull/11470#issuecomment-215332039

jreback commented:
>>>In [5]: pd.to_datetime([1,2],unit='D',origin='julian')
ValueError: cannot convert input with unit: D
@jreback why is 'unit D' not valid? You get a out of bounds date here, but that is not really clear from that message.

No this is exactly what I want to catch (then re-raise) but with a better error message. That is the underlying message propogating up.

jorisvandenbossche commented:
On my comment above, my idea is to only allow `origin` when also `unit` is used/when the argument are integers (as I don't think other options make any sense, like `pd.to_datetime('2012-01-01', origin='julian')` makes no sense, so better to just disallow). 
But I think your changes make that easier now.

jorisvandenbossche commented:
> No this is exactly what I want to catch (then re-raise) but with a better error message. That is the underlying message propogating up.

Not fully clear. Is the error message "cannot convert input with unit: D" the one you want to raise to the user, or the one you want to catch to re-raise with a better message?

jreback commented:
So the issue with 

```In [3]: pd.to_datetime([1,2],origin='julian')```

is its doing this computation.
```
In [8]: np.array([1, 2])-pd.Timestamp(0).to_julian_date()
Out[8]: array([-2440586.5, -2440585.5])
```

Which I am not sure what this should be

jreback commented:
@jorisvandenbossche that is the one I want to catch (not this is ONLY for passing ``origin='julian'`` case)

jreback commented:
```
In [14]: pd.to_datetime(pd.Timestamp(0).to_julian_date(),origin='julian',unit='D')
Out[14]: Timestamp('1970-01-01 00:00:00')
```

jreback commented:
let me push the bug I found independtly. Then this can be revisited.

jreback commented:
xref #13052

sumitbinnani commented:
Was a bit occupied this weekend and seems I missed on a lot. This is what I need to submit:

1. Add testcases for OverflowError
2. Add check and testcases for combination of arg!=int and origin!=''unix"
3. Re-raise OutOfBoundsDatetime for ValueError

Correct me if I am going wrong somewhere.

jreback commented:
@sumitbinnani ``OverflowError`` is no longer, its just a ``ValueError`` now. but yes, need some test cases.

otherwise yes.

jreback commented:
@sumitbinnani can you rebase / update & address those issues as discussed. want to get this in as soon as possible.

sumitbinnani commented:
Would submit the patch by the end of week. Sorry for the delay :-(

sumitbinnani commented:
Facing a bit difficulty implementing check for ``origin`` and ``arg`` as suggested by @jorisvandenbossche. Any suggestions? @jreback?

jreback commented:
@sumitbinnani can you show a concise example

Winand commented:
(haven't checked this)
```
if origin != 'epoch':
    if not isinstance(arg, (int, np.integer)):
        raise ValueError("arg must be integer when origin is used")
    if origin == 'julian':
        if unit != 'D':
            raise ValueError("unit must be 'D' for origin='julian'")
        arg = arg - tslib.Timestamp(0).to_julian_date()
    else:
        offset = tslib.Timestamp(origin) - tslib.Timestamp(0)
```

jreback commented:
@sumitbinnani can you rebase / update

jreback commented:
@sumitbinnani can you update

sumitbinnani commented:
@jreback: Sorry for the delay. I have been running busy, and am a bit occupied till July 20. Would make all the required changes by July 25. Once again sorry for the late communication.

jreback commented:
@sumitbinnani can you rebase and update

jreback commented:
@sumitbinnani if you'd like to rebase / update

sumitbinnani commented:
@jreback: I have rebased the local branch but a few changes are pending. Owing to a bit gusy schedule, would make necessary changes starting July 20, and we would probably be able to close the issue by 25th. Sorry for the delay.

jorisvandenbossche commented:
@sumitbinnani If you would have time to update this, very welcome

sumitbinnani commented:
Almost done. Left with a few merge conflicts.

sumitbinnani committed:
ENH: Added origin parameter to to_datetime

sumitbinnani committed:
TST: Added tests for origin paramters

sumitbinnani committed:
DOC: Update whatsnew in v0.20.0 to include origin parameter

sumitbinnani commented:
I hope all the cases have been accounted, @jorisvandenbossche @jreback?

sumitbinnani committed:
CLN: Adhering to pep8 guidelines

sumitbinnani committed:
CLN: Code Cleanup

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -20,7 +20,7 @@ Check the :ref:`API Changes <whatsnew_0200.api_breaking>` and :ref:`deprecations
 New features
 ~~~~~~~~~~~~
-
+- ``pd.to_datetime`` has a new parameter, ``origin``, to define an offset for ``DatetimeIndex`` (:issue:`11276`, :issue:`11745`)
```
move to 0.19.0. pls add a sub-section with an example (you can use same one in doc-string). in the enhancements section.

jreback commented on pandas/tseries/tools.py:
```diff
@@ -238,6 +238,17 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
+        default 'epoch'.
+        Define relative offset for the returned dates.
+
+        - If 'epoch', offset is set to 1-1-1970.
```
write this in ISO: (1970-01-01)

jreback commented on pandas/tseries/tools.py:
```diff
@@ -238,6 +238,17 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
+
+        - If 'epoch', offset is set to 1-1-1970.
+        - If 'julian', unit must be 'D', and offset is set to beginning of
+          Julian Calendar.
```
indicate what date this is

jreback commented on pandas/tseries/tools.py:
```diff
@@ -238,6 +238,17 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
+        - If Timestamp convertible, offset is set to Timestamp identified by
+          origin.
+
+        .. versionadded: 0.20.0
```
0.19.0

jreback commented on pandas/tseries/tools.py:
```diff
@@ -294,8 +305,14 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
+    1    1960-01-02
+    ...
+    98   1960-04-08
+    99   1960-04-09
```
add a comment that these are using non-epoch origins

sumitbinnani replied:
Done.

jreback commented on pandas/tseries/tools.py:
```diff
@@ -406,22 +423,45 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
+        try:
+            offset = tslib.Timestamp(origin) - tslib.Timestamp(0)
+        except ValueError:
+            if errors == 'raise':
```
I don't think this is what @jorisvandenbossche meant earlier. If the ``offset`` is unconvertible, this is a straight ``ValuError``, the errors parameters does not matter. The errors applies to the arg.

sumitbinnani replied:
I assumed that the errors message is to be suppressed in case the function is not able to parse the input. Should I simply raise the error, @jorisvandenbossche?

shoyer replied:
Indeed, we only want to raise an error for invalid `arg`, not `origin`.

sumitbinnani committed:
CLN: Worked on Code Reviews

sumitbinnani committed:
ENH: Added Error Handling for invalid Origin

sumitbinnani commented:
Have updated the code to raise error for invalid origin. Also modified the test cases accordingly.

jreback commented on doc/source/whatsnew/v0.19.0.txt:
```diff
@@ -21,6 +21,7 @@ Highlights include:
 
 New features
 ~~~~~~~~~~~~
+- ``pd.to_datetime`` has a new parameter, ``origin``, to define an offset for ``DatetimeIndex`` (:issue:`11276`, :issue:`11745`)
```
need a mini-example here

jreback replied:
put in a separate sub-section

sumitbinnani replied:
Done!

jreback commented:
@sumitbinnani pls address the other comments

jorisvandenbossche commented on pandas/tseries/tests/test_timeseries.py:
```diff
@@ -772,6 +772,48 @@ def test_to_datetime_unit(self):
+        julian_dates = pd.date_range(
+            '2014-1-1', periods=10).to_julian_date().values
+        result = Series(pd.to_datetime(
+            julian_dates, unit='D', origin='julian'))
```
You can omit the `Series( ..)` here and in the 'expected' below as well. It will then return indexes (so need to use `assert_index_equal` instead of series), but it will make the test a little bit easier to follow

sumitbinnani replied:
using assert_index_equal raises AssertionError.
```
AssertionError: [index] Expected type <class 'pandas.indexes.base.Index'>, found
```

jorisvandenbossche replied:
ah, yes, sorry. It then probably returns a numpy array and not an Index. Then you can leave it as is.

jorisvandenbossche commented on pandas/tseries/tools.py:
```diff
@@ -238,6 +238,18 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
         speed by ~5-10x.
+    origin : scalar convertible to Timestamp / string ('julian', 'epoch'),
+        default 'epoch'.
+        Define relative offset for the returned dates.
```
"returned dates" -> isn't the offset to interpret the input? (as the returned dates should always be correct dates, not with an offset)

Also, can you specify this is only relevant when converting integers (based on the `unit` kwarg)?

sumitbinnani replied:
Can you elaborate a bit? I am not able to understand.

jorisvandenbossche replied:
I think that "Define relative offset for the returned dates" is not fully correct. As the offset you specify with `origin` is used to correctly parse the *input*, not the *returned dates*

So I would maybe replace it with something like

```
Date to parse numeric values as number of units (determined by `unit`) since this origin.
```

Does this sounds understandable to you? (english is also not my native tongue)

sumitbinnani replied:
Done.

jorisvandenbossche commented on pandas/tseries/tools.py:
```diff
@@ -294,8 +306,16 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
-    """
+    Using non-epoch origins to parse date
+
+    >>> pd.to_datetime([1,2,3], unit='D', origin=Timestamp('1960-01-01'))
```
`Timestamp` -> `pd.Timestamp`

sumitbinnani replied:
Done.

jorisvandenbossche commented on pandas/tseries/tools.py:
```diff
@@ -406,22 +426,40 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
+        try:
+            offset = tslib.Timestamp(origin) - tslib.Timestamp(0)
+        except ValueError:
+            raise ValueError("Invalid Origin or Origin Out of Bound")
```
use lower case, and maybe use `'origing'` to indicate it is a kwarg

sumitbinnani replied:
` ``'origing'`` to indicate it is a kwarg` What does this mean?

jorisvandenbossche replied:
sorry the 'g' was a typo :)

I meant converting `Origin` to `'origin'` (so lower case and quotes)

sumitbinnani replied:
Done.

jorisvandenbossche commented:
@sumitbinnani Do you have time to update for the latest comments?

sumitbinnani committed:
CLN: Addressing Code Reveiws

sumitbinnani commented:
Pertaining to check when ``origin='julian'``, how can I check whether the ``arg`` is an array of ``integer`` or ``float``?

_# usually busy during weekdays and hence the delay :-(_

jorisvandenbossche commented on pandas/tseries/tools.py:
```diff
@@ -294,8 +306,16 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
+    0    1960-01-01
+    1    1960-01-02
+    ...
+    98   1960-04-08
```
The output here (Series of lenght 100) does not correspond to the input (length of 3)

jorisvandenbossche commented on pandas/tseries/tools.py:
```diff
@@ -238,6 +238,18 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
+        default 'epoch'.
+        Define relative offset for the returned dates.
+
+        - If 'epoch', offset is set to 1970-01-01.
```
I would maybe use `origin` instead of `offset` (here and in lines below). The keyword is named like that, and I also think it is more correct word to use. And offset is something like a 'timedelta', while the origins here are actual 'timestamps'

sumitbinnani replied:
Done.

jorisvandenbossche commented:
> Pertaining to check when origin='julian', how can I check whether the arg is an array of integer or float?

Maybe this check is not needed. As there is already an informative message if you pass non-numeric values when `unit` is specified:

```
In [38]: pd.to_datetime(['2012-01-01'], unit='D')
...
ValueError: non convertible value 2012-01-01 with the unit 'D'
```

So I would just make sure that `unit` is not None if `origin` is specified.

> # usually busy during weekdays and hence the delay :-(

No problem! We will just keep pinging you from time to time :-)

sumitbinnani committed:
CLN: Worked on Code Review.

sumitbinnani commented:
does it look good, @jreback, @jorisvandenbossche?

sumitbinnani commented:
does it look good, @jreback, @jorisvandenbossche?

jorisvandenbossche commented:
Another issues: When trying to convert the number of days since year 1, you get:
```
In [13]: pd.to_datetime([2000*365], unit='D', origin='0001-01-01')
...
ValueError: Invalid 'origin' or 'origin' Out of Bound
```

so the origin itself is out of bound (even when the actual date (if converted) is not). That seems to limit the use of custom origins a bit. Could we solve this? 
Problem is how to parse the string ...

@jreback Are there precedents in pandas where we try to deal with out of bounds datetimes?

jorisvandenbossche commented on pandas/tseries/tools.py:
```diff
@@ -294,8 +307,16 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
+    0    1960-01-01
+    1    1960-01-02
+    ...
+    98   1960-04-08
```
The output here of the example does not yet match the example code

sumitbinnani replied:
Opps. My Bad. Updated the documentation with proper output.

sumitbinnani committed:
DOC: Updated Documentation for origin parameter

Rectified mistake in documentation.

sumitbinnani commented:
Have rectified the mistake in example. Unsure about how to proceed for the aforementioned invalid-origin :-(

jreback commented:
```
In [13]: pd.to_datetime([2000*365], unit='D', origin='0001-01-01')
...
ValueError: Invalid 'origin' or 'origin' Out of Bound
```

I think this is fine for now. We *could* make this happen, but the code then gets convoluted, but it is so rare as not to be important.

I would just assert that this example raises.

jreback commented on doc/source/whatsnew/v0.19.0.txt:
```diff
@@ -24,6 +24,16 @@ New features
 
 .. _whatsnew_0190.dev_api:
 
+to_datetime can be used with Offset
```
``to_datetime has gained an origin kwarg``.

don't call this ``Offset`` which is a very specific meaning

jreback replied:
in the doc-string this is a 'reference date', so I would use that here.

sumitbinnani replied:
Done.

jreback commented on doc/source/whatsnew/v0.19.0.txt:
```diff
@@ -24,6 +24,16 @@ New features
 
+to_datetime can be used with Offset
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+``pd.to_datetime`` has a new parameter, ``origin``, to define an offset for ``DatetimeIndex``.
```
say its a starting offset

sumitbinnani replied:
Done.

jreback commented on pandas/tseries/tools.py:
```diff
@@ -406,22 +425,39 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
-        return _convert_listlike(arg, box, format, name=arg.name)
-    elif is_list_like(arg):
-        return _convert_listlike(arg, box, format)
+    def result_without_offset(arg):
```
``offset`` -> ``origin``

sumitbinnani replied:
Done.

jreback commented on pandas/tseries/tools.py:
```diff
@@ -406,22 +425,39 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
+        if origin == 'julian':
+            if unit != 'D':
+                raise ValueError("unit must be 'D' for origin='julian'")
+            arg = arg - tslib.Timestamp(0).to_julian_date()
```
``arg`` needs validation here, it MUST be an integer

sumitbinnani replied:
not sure how to proceed with the aforementioned check.

jreback commented on pandas/tseries/tools.py:
```diff
@@ -406,22 +425,39 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
+    result = result_without_offset(arg)
+
+    offset = None
+    if origin != 'epoch' and origin != 'julian':
```
``if origin not in ['epoch', 'julian']:``

sumitbinnani replied:
Done.

jreback commented on pandas/tseries/tests/test_timeseries.py:
```diff
@@ -772,6 +772,48 @@ def test_to_datetime_unit(self):
+            julian_dates, unit='D', origin='julian'))
+        expected = Series(pd.to_datetime(
+            julian_dates - pd.Timestamp(0).to_julian_date(), unit='D'))
+        assert_series_equal(result, expected)
```
add tests for invalid when ``origin='julian'``. IOW non-integer, out-of-range julian dates.

jreback commented:
can you rebase / update?

sumitbinnani committed:
CLN: Worked on Code Reveiew

jorisvandenbossche commented:
@sumitbinnani Can you move the whatsnew notice to the v0.20.0.txt file? As 0.19.0 has been releases in the meantime (also, all the other changes in the 0.19.0.txt file should not be included in here)

sumitbinnani committed:
DOC: Adding doc for new kwarg origin

sumitbinnani commented:
@jorisvandenbossche The v0.19.0 had been reverted to the master. Changes added to v0.20.0.

jorisvandenbossche commented:
Some examples that now go wrong:

```
In [1]: pd.to_datetime("2005-01-01", origin="1960-01-01")
Out[1]: Timestamp('1995-01-01 00:00:00')

In [2]: pd.to_datetime(1, origin="julian", unit='D')
...
OutOfBoundsDatetime: cannot convert input with unit 'D'

In [3]: pd.to_datetime(2456658, origin='julian', unit='D')
Out[3]: Timestamp('2013-12-31 00:00:00')

In [4]: pd.to_datetime(2456658, origin='0001-01-01', unit='D')
...
OutOfBoundsDatetime: cannot convert input with unit 'D'
```

So in `In [1]`, this should raise, as the origin should *only* be used when having numeric input. `In [2]` is correct I think, but we should add a test for this that it indeed raises an OutOfBoundsDatetime error. `In [4]` should work (`In [3]` is just to show this is a number that is not out of bounds).

jorisvandenbossche commented:
@sumitbinnani what I think you could do to simplify the code (having all handling of origin in one place) is the following: start with processing the `origin` keyword (and only do this for numerical input), convert it to an offset (according the the `unit`), and adapt the numerical input with this offset *before* further parsing it.

So something like (pseudocode)

```
if origin != 'epoch':
    if origin == 'julian':
        ... (what you already have in this case)
        arg = arg - tslib.Timestamp(0).to_julian_date()
    else:
        ... (parse origin and convert to correct number according to unit)
        arg = arg - offset

... (original code to parse `arg` (what you now put in the 'intermediate_result' function)
```

A way to deal with the out of bounds origin would be to use `np.datetime64(dateutil.parser.parse(origin))`, but it's OK to leave that for now.

jreback commented:
can you rebase / update according to comments

jreback commented:
@sumitbinnani I know this has been open a long time. can you rebase and we'll see if can get it in?

pls change release note to 0.20.0

sumitbinnani commented:
@jreback There are quite a many changes. I would start clean and update asap. Sorry for the delay.

bashtage commented:
@sumitbinnani Useful feature so would be good to get this in

bashtage commented on pandas/tseries/tools.py:
```diff
@@ -406,22 +425,43 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
-        return _convert_listlike(arg, box, format, name=arg.name)
-    elif is_list_like(arg):
-        return _convert_listlike(arg, box, format)
+    def intermediate_result(arg):
```
Is there some reason this is a function?  It looks like it should just be an `if, else if, else`

bashtage commented on pandas/tseries/tools.py:
```diff
@@ -238,6 +238,19 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
+        Define reference date. The numeric values would be parsed as number
+        of units (defined by `unit`) since this reference date.
+
+        - If 'epoch', origin is set to 1970-01-01.
```
You could consider other standard offsets: `excel`, Jan 0 1900, `stata` Jan 0, 1960, `matlab`, Jan 0, 0000?  Just an idea

jreback commented:
superseded by #15828 

@sumitbinnani I rebase and updated your PR. thanks!
